### PR TITLE
Add ReflectStatus ResourceInterperter interface to grab workload status

### PR DESCRIPTION
**Pull request by XiShanYongYe-Chang**
Signed-off-by: changzhen <changzhen5@huawei.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Add ReflectStatus ResourceInterperter interface to grab workload status, so as to support granularity control of workload status collection.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

The default implementation of some resources kind(such as `deployment`) is not added. 

Add default reflectstatus implementation with:
- Deployment kind
- Service kind
- Ingress kind
- Job kind
- DaemonSet kind
- StatefulSet kind

**Does this PR introduce a user-facing change?**:

```release-note
Supports granularity control of workload status collection with InterpretStatus InterpreterOperation.
```

### CODE REVIEW

File: pkg/resourceinterpreter/defaultinterpreter/default.go [hunk 1]
```diff
@@ -106,3 +112,23 @@ func (e *DefaultInterpreter) GetDependencies(object *unstructured.Unstructured)
...
+
+// ReflectStatus returns the cluster object's running status after the grab.
+func (e *DefaultInterpreter) ReflectStatus(object *unstructured.Unstructured) (status *runtime.RawExtension, err error) {
+	handler, exist := e.reflectStatusHandlers[object.GroupVersionKind()]
+	if exist {
+		return handler(object)
+	}
+
+	// for no default build-in resource kind, directly collect the entire status.
+	statusMap, _, err := unstructured.NestedMap(object.Object, "status")
```
Comment (RainbowMango) [inline pkg/resourceinterpreter/defaultinterpreter/default.go:124]: Please add a unit test for this function. I want to see what happened if no `.status` field.
Maybe we can use the second return value to suppress error logs.

Comment (XiShanYongYe-Chang) [reply pkg/resourceinterpreter/defaultinterpreter/default.go]: Done.

File: pkg/util/helper/workstatus.go [hunk 2]
```diff
@@ -283,7 +283,7 @@ func IsWorkContains(workStatus *workv1alpha1.WorkStatus, targetResource schema.G
 }
 
 // BuildStatusRawExtension builds raw JSON by a status map.
-func BuildStatusRawExtension(status map[string]interface{}) (*runtime.RawExtension, error) {
+func BuildStatusRawExtension(status interface{}) (*runtime.RawExtension, error) {
```
Comment (RainbowMango) [inline pkg/util/helper/workstatus.go:286]: Just a question, why change the parameter here?

File: pkg/resourceinterpreter/customizedinterpreter/customized.go [hunk 3]
```diff
@@ -293,3 +294,18 @@ func (e *CustomizedInterpreter) GetDependencies(ctx context.Context, attributes
 
 	return response.Dependencies, matched, nil
 }
+
+// ReflectStatus returns the cluster object's running status after the grab.
```
Comment (RainbowMango) [inline pkg/resourceinterpreter/customizedinterpreter/customized.go:298]: Update the comments.

File: pkg/controllers/status/workstatus_controller.go [hunk 4]
```diff
@@ -275,10 +277,10 @@ func (c *WorkStatusController) recreateResourceIfNeeded(work *workv1alpha1.Work,
 
 // reflectStatus grabs cluster object's running status then updates to its owner object(Work).
 func (c *WorkStatusController) reflectStatus(work *workv1alpha1.Work, clusterObj *unstructured.Unstructured) error {
-	statusMap, _, err := unstructured.NestedMap(clusterObj.Object, "status")
+	statusRaw, err := c.ResourceInterpreter.ReflectStatus(clusterObj)
```
Comment (RainbowMango) [inline pkg/controllers/status/workstatus_controller.go:280]: If `statusRaw` == nil and the status in `Work` is nil too, then we can save one update operation.

There is a corner case that, the status in `Work` is not nil, but this time we get `nil` from the interpreter, then, we have to update the `Work`.

File: pkg/resourceinterpreter/defaultinterpreter/reflectstatus.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+func reflectDeploymentStatus(object *unstructured.Unstructured) (*runtime.RawExtension, error) {
+	statusMap, exist, err := unstructured.NestedMap(object.Object, "status")
+	if err != nil {
+		klog.Errorf("Failed to get status field from %s(%s/%s), error: %v",
+			object.GetKind(), object.GetNamespace(), object.GetName(), err)
+		return nil, err
+	}
+	if !exist {
+		klog.Warningf("Obtain no status from %s(%s/%s).",
```
Comment (RainbowMango) [inline pkg/resourceinterpreter/defaultinterpreter/reflectstatus.go:32]:
```suggestion
		klog.Errorf("Failed to grab status from xxx which should have status field.",
```

File: pkg/util/helper/workstatus.go (hunk 2 above)
Comment (XiShanYongYe-Chang) [reply pkg/util/helper/workstatus.go:286]: So we can directly use this function when the input parameter is `appsv1.DeploymentStatus` or something else.

In fact, we will marshal the status inside this method, the input parameter of the `json.Marshal` is not restrict with `map[string]interface{}` type.

This change can makes this function more general.

File: pkg/controllers/status/workstatus_controller.go (hunk 4 above)
Comment (XiShanYongYe-Chang) [reply pkg/controllers/status/workstatus_controller.go:280]: How about adding an equal judgment when updating.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**